packages/contents/src/drivefs.ts

```diff
@@ -0,0 +1,517 @@
+  514 /*O_RDWR|O_TRUNC*/: true,
+  577 /*O_WRONLY|O_CREAT|O_TRUNC*/: true,
+  578 /*O_CREAT|O_RDWR|O_TRUNC*/: true,
+  705 /*O_WRONLY|O_CREAT|O_EXCL|O_TRUNC*/: true,
+  706 /*O_RDWR|O_CREAT|O_EXCL|O_TRUNC*/: true,
+  1024 /*O_APPEND*/: true,
+  1025 /*O_WRONLY|O_APPEND*/: true,
+  1026 /*O_RDWR|O_APPEND*/: true,
+  1089 /*O_WRONLY|O_CREAT|O_APPEND*/: true,
+  1090 /*O_RDWR|O_CREAT|O_APPEND*/: true,
+  1153 /*O_WRONLY|O_EXCL|O_APPEND*/: true,
+  1154 /*O_RDWR|O_EXCL|O_APPEND*/: true,
+  1217 /*O_WRONLY|O_CREAT|O_EXCL|O_APPEND*/: true,
+  1218 /*O_RDWR|O_CREAT|O_EXCL|O_APPEND*/: true,
+  4096 /*O_RDONLY|O_DSYNC*/: true,
+  4098 /*O_RDWR|O_DSYNC*/: true,
+};
+
+// Types and implementation inspired from
+// https://github.com/jvilk/BrowserFS
+// https://github.com/jvilk/BrowserFS/blob/a96aa2d417995dac7d376987839bc4e95e218e06/src/generic/emscripten_fs.ts
+// And from https://github.com/gzuidhof/starboard-notebook
+export interface IStats {
+  dev: number;
+  ino: number;
+  mode: number;
+  nlink: number;
+  uid: number;
+  gid: number;
+  rdev: number;
+  size: number;
+  blksize: number;
+  blocks: number;
+  atime: Date;
+  mtime: Date;
+  ctime: Date;
+  timestamp?: number;
+}
+
+export interface IEmscriptenFSNode {
+  name: string;
+  mode: number;
+  parent: IEmscriptenFSNode;
+  mount: { opts: { root: string } };
+  stream_ops: IEmscriptenStreamOps;
+  node_ops: IEmscriptenNodeOps;
+  timestamp: number;
+}
+
+export interface IEmscriptenStream {
+  node: IEmscriptenFSNode;
+  nfd: any;
+  flags: string;
+  position: number;
+  file?: DriveFS.IFile;
+}
+
+export interface IEmscriptenNodeOps {
+  getattr(node: IEmscriptenFSNode): IStats;
+  setattr(node: IEmscriptenFSNode, attr: IStats): void;
+  lookup(parent: IEmscriptenFSNode, name: string): IEmscriptenFSNode;
+  mknod(
+    parent: IEmscriptenFSNode,
+    name: string,
+    mode: number,
+    dev: any
+  ): IEmscriptenFSNode;
+  rename(oldNode: IEmscriptenFSNode, newDir: IEmscriptenFSNode, newName: string): void;
+  unlink(parent: IEmscriptenFSNode, name: string): void;
+  rmdir(parent: IEmscriptenFSNode, name: string): void;
+  readdir(node: IEmscriptenFSNode): string[];
+  symlink(parent: IEmscriptenFSNode, newName: string, oldPath: string): void;
+  readlink(node: IEmscriptenFSNode): string;
+}
+
+export interface IEmscriptenStreamOps {
+  open(stream: IEmscriptenStream): void;
+  close(stream: IEmscriptenStream): void;
+  read(
+    stream: IEmscriptenStream,
+    buffer: Uint8Array,
+    offset: number,
+    length: number,
+    position: number
+  ): number;
+  write(
+    stream: IEmscriptenStream,
+    buffer: Uint8Array,
+    offset: number,
+    length: number,
+    position: number
+  ): number;
+  llseek(stream: IEmscriptenStream, offset: number, whence: number): number;
+}
+
+export class DriveFSEmscriptenStreamOps implements IEmscriptenStreamOps {
+  private fs: DriveFS;
+
+  constructor(fs: DriveFS) {
+    this.fs = fs;
+  }
+
+  public open(stream: IEmscriptenStream): void {
```

Comment:
Maybe we could drop the `public` here (and below) for consistency with the rest of the code base:

```suggestion
  open(stream: IEmscriptenStream): void {
```